<a href="https://colab.research.google.com/github/Himeshverma/Algorithm-and-Problem-Solving/blob/main/Lab_2_AlgorithmsBasics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# LAB_2. Algorithm's Basics

#1.1 Problem Definition
This file contains a series of programming exercises designed for practice and understanding of recursion, time complexity, and array/list manipulation task. Your task is to: 1. Write in the form of comment what each function does. 2. Calculate the time complexity of each function. A Sample example is given in Exercie 1

In [ ]:
# Exercise 1
def func1(a, b):
  result = 0
  for i in range(0, b):
    result += a
  return result

In [ ]:
# Sample function call: func1(5, 3)
x=print(func1(4,3))
#the function performs multiplication using repeated addition

12


In [ ]:
# Exercise 2
def func2(a, b):
  result = 1
  while b > 0:
    result = result * a
    b = b - 1
  return result

In [ ]:
# Sample function call: func2(2, 4)

In [ ]:
# Exercise 3
def func3(num):
   result = 0
   while num > 0:
    result = result + num % 10
    num //= 10
    return result

In [ ]:
# Sample function call: func3(1234)

In [ ]:
# Exercise 4
def func4(a, b):
    b.sort() # Time complexity for this function is O(Nlog N) where N = len(b)
    intersect = 0
    for x in a:
        if binary_search(b, x) >= 0:
            intersect = intersect + 1
    return intersect

def binary_search(arr, target):
    low, high = 0, len(arr) - 1
    while low <= high:
        mid = (low + high) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            low = mid + 1
        else:
            high = mid - 1
    return -1

In [ ]:
# Sample function call: func4([1, 2, 3, 4], [2, 4, 6, 8])

In [ ]:
# Exercise 5
def func5(n):
    guess = 1
    while guess * guess <= n:
        if guess * guess == n:
            return guess
        guess = guess + 1
    return None

In [ ]:
# Sample function call: func5(16)

In [ ]:
# Exercise 6
def func8(A):
    func8_helper(A, 0, len(A) -1)

def func8_helper(A, i, j):
    #Input: An array A and two indices i and j.
    n = j - 1 + 1
    if n ==2:
        if A[i] > A[j]:
            # Swap A[i] and A[j]
            A[i], A[j] = A[j], A[i]
        elif n > 2:
            m = n // 3
            func8_helper(A, i, j - m)
            func8_helper(A, i + m, j)
            func8_helper(A, i, j - m)
        return A


In [ ]:
#Sample function call: func8([3,1,2,7,8,6])

In [ ]:
# Exercise 7
def func12(arr, k):
    arr_len = len(arr)
    k = k % arr_len
    # Initialize a new array of the same length as 'arr'
    custom_arr = [0] * arr_len

    for i in range(arr_len):
        mystery_index = (i + k) % arr_len
        custom_arr[mystery_index] = arr[i]
    return custom_arr



In [ ]:
#Sample function call: func12([1, 2, 3, 4, 5], 2)


##1.2 Arrays
Programs on 1-d arrays like - sum of elements of array, searching an element inarray, finding minimum and maximum element in array, count the number ofeven and odd numbers in array. For all such programs, also find the timeComplexity, compare if there are multiple methods

In [15]:
# write a function named as sum_arr to find sum of elements in given array
def sum_arr(arr):
  # This function calculates the sum of all elements in a 1-D array.
  # Time complexity: O(N), where N is the number of elements in the array.
  total_sum = 0
  for element in arr:
    total_sum += element
  return total_sum

In [16]:
#call the function
#arr = [12, 3, 4, 15,45,78,450]
#print(sum_arr)
arr = [12, 1234, 45, 67, 1]
print(sum_arr(arr))

1359


In [20]:
def max_min(arr):

  if not arr:
    return None, None # Return None for both if the array is empty

  min_val = arr[0]
  max_val = arr[0]

  for element in arr:
    if element < min_val:
      min_val = element
    if element > max_val:
      max_val = element
  return min_val, max_val

**Reasoning**:
To test the `max_min` function, I will define a sample array and then call `max_min` with that array, printing the returned minimum and maximum values.

In [21]:
# Sample function call: max_min([12, 1234, 45, 67, 1])
arr_for_minmax = [12, 1234, 45, 67, 1, -5, 2000]
min_val, max_val = max_min(arr_for_minmax)
print(f"Array: {arr_for_minmax}")
print(f"Minimum element: {min_val}")
print(f"Maximum element: {max_val}")

Array: [12, 1234, 45, 67, 1, -5, 2000]
Minimum element: -5
Maximum element: 2000


In [22]:
# call sum_arr to find sum of elements in given array for arr = [12, 1234, 45, 67, 1]
arr = [12, 1234, 45, 67, 1]
print(sum_arr(arr))

1359


In [24]:
# write a function named as Counting_EvenOdd to count number of even and odd elements in an array
def Counting_EvenOdd(arr):
  even_count = 0
  odd_count = 0
  for element in arr:
    if element % 2 == 0:
      even_count += 1
    else:
      odd_count += 1
  return even_count, odd_count # Add a return statement to output the counts

In [25]:
# call above function for arr = [2, 3, 4, 5, 6,7,8,9,10,11,12,13]
arr = [2, 3, 4, 5, 6,7,8,9,10,11,12,13]
print(Counting_EvenOdd(arr))

(6, 6)


In [26]:
# write a function named as linear_Search to perform a linear search in 1D-array
def linear_Search(arr, target):
  for i in range(len(arr)):
    if arr[i] == target:
      return i  # Return the index if the target is found
  return -1


In [29]:
# call the function liner_Search for array = [1, 3, 5, 7, 9]
arr = [1, 3, 5, 7, 9]
target = 5
print(linear_Search(arr, target))

2


In [31]:
# write a function named as Binary_Search to perform binary search in 1D-array
def Binary_Search(arr, target):
  low = 0
  high = len(arr) - 1
  while low <= high:
    mid = (low + high) // 2
    if arr[mid] == target:
      return mid
    elif arr[mid] < target:
      low = mid + 1
    else:
      high = mid - 1
  return -1 # Return -1 if the target is not found

In [36]:
# call the function Binary_Search for array
arr = [1, 3, 5, 7, 9,15,16,14,45]
target = 9
print(Binary_Search(arr, target))


4


# Task
Implement Python functions for 1-D array operations: `sum_arr` (sum of elements), `max_min` (minimum and maximum elements), `Counting_EvenOdd` (count even/odd elements), `linear_Search` (linear search), and `Binary_Search` (binary search for sorted arrays). Each function should include comments explaining its purpose and time complexity. After implementation, call and test each function with the following sample arrays: `sum_arr` with `[12, 1234, 45, 67, 1]`; `Counting_EvenOdd` with `[2, 3, 4, 5, 6,7,8,9,10,11,12,13]`; and `linear_Search` and `Binary_Search` with `[1, 3, 5, 7, 9]` (ensure `Binary_Search` is used on a sorted array). Finally, summarize the implemented array functions, their time complexities, and compare linear and binary search methods.

## Implement Sum of Array Elements

### Subtask:
Write a Python function `sum_arr` to calculate the sum of all elements in a given 1-D array. Include comments explaining the function's purpose and its time complexity.


**Reasoning**:
Implement the `sum_arr` function as per the instructions, including comments for its purpose and time complexity, in the designated code cell.



# Task
The sum of elements for `sum_arr` with the array `[12, 1234, 45, 67, 1]` has been calculated.

## Call sum_arr

### Subtask:
Call the `sum_arr` function with the array `[12, 1234, 45, 67, 1]` and print the returned sum.


**Reasoning**:
The subtask requires calling the `sum_arr` function with a specified array and printing the result. I will create a code cell to define the array, call the function, and print its output.



## Final Task

### Subtask:
Indicate that the sum calculation for `sum_arr` is complete.


## Summary:

### Data Analysis Key Findings
*   The sum of the elements in the array `[12, 1234, 45, 67, 1]` was calculated to be 1359.

### Insights or Next Steps
*   The `sum_arr` function successfully calculated the sum of the provided array.
*   The task of calculating the sum for `sum_arr` is complete.
